### PR TITLE
Support more HTML elements and CSS attributes and fix CSS filter crashes

### PR DESCRIPTION
**Pull request by torusrxxx**
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdi
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/dir

### CODE REVIEW

**Review by bertm — CHANGES_REQUESTED** [review]
This looks reasonable but please do add a test for this. The bare minimum would be something like / an extension to what is introduced in
8d12ea0dfa8d50c462c9e6584617cbfd13792d83

**Comment by torusrxxx** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figure
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/mark
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ruby
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/rp
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/rt

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:1627]: The shorthands are deprecated except for SVG (see https://developer.mozilla.org/en-US/docs/Web/CSS/writing-mode ), so I think we should avoid them.

Supporting SVG as content type would be a huge project, because that can embed far too many ways to access remote content.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:218]: no, should not be blocked: this applies to any link. As long as visited is stripped out, there’s no privacy risk.

⇒ please adjust the comment.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:238]: is dir necessary here? We already have a dedicated parsing rule for it. If it is: please add a test that fails without your change.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:231]: this is the solution to the link/visited privacy problems.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:230]: target looks dangerous (it checks which link is used to access a page). It can be used to trigger background fetches when navigating (just verified: the network request is sent when the target is clicked), so it may be necessary to restrict its content to non-links to avoid timing attacks.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:233]: we do not have custom element support (it requires javascript), so defined does not make sense. Please remove it.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:239]: also only used for forms.

File: src/freenet/client/filter/FilterUtils.java
Comment (ArneBab) [inline src/freenet/client/filter/FilterUtils.java:380]: good catch!

File: test/freenet/client/filter/FilterUtilsTest.java
Comment (ArneBab) [inline test/freenet/client/filter/FilterUtilsTest.java:67]: can you add more color-tests? switching to regexes usually needs tests for edge-cases.

File: src/freenet/client/filter/HTMLFilter.java
Comment (ArneBab) [inline src/freenet/client/filter/HTMLFilter.java:1006]: which elements did you add? the reordering makes that hard to figure out.

**Comment by ArneBab** [comment]
This is a neat addition, but the added elements still need tests. I know that these can be annoying to write, since they feel obvious, but the last two times I omitted them and Bombe reminded me with an ENOTEST, one of those would have actually shown a bug (that luckily got found before merging when I added the test Bombe had requested).

⇒ please do test. These filters are critical for safe operation in browsers that may query into the clearnet on errors here.

File: src/freenet/client/filter/ElementInfo.java
Comment (torusrxxx) [reply src/freenet/client/filter/ElementInfo.java:234]: Why do we need to discuss where these can be used? As long as it's valid it should be here, right? And if it has problems, we can add it into the BANNED_PSEUDOCLASS. Freesites do have forms, could be Library, commentmaker or some other future plugin.

File: src/freenet/client/filter/ElementInfo.java
Comment (torusrxxx) [reply src/freenet/client/filter/ElementInfo.java:233]: Remove it or add it into the BANNED_PSEUDOCLASS?

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java:1627]: I think these are valid and not dangerous so should be here. Freenet also has lots of other deprecated things, but still allowed because it's safe.

File: src/freenet/client/filter/ElementInfo.java
Comment (torusrxxx) [reply src/freenet/client/filter/ElementInfo.java:230]: target is usually used to highlight a section heading. Clicking different links should already trigger different network effects because they can link to different pages, yeah.

File: src/freenet/client/filter/HTMLFilter.java
Comment (torusrxxx) [reply src/freenet/client/filter/HTMLFilter.java:1006]: "bdi","figcaption","figure","main","mark","rp","rt","ruby". I reordered them because it's hard to find which element is still missing. All these new elements are added to tests HTML5_TAGS and HTML5_BDI_RUBY.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:233]: BANNED_PSEUDOCLASS is better: CSSTokenizerFilter::HTMLElementVerifier defines that as `banned (but otherwise valid) selector`.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:234]: I’m being careful with adding selectors that might not be usable, because I have to check with each element, whether it could allow sneaking in an unfiltered URL. That’s why the review of these takes a while (need to look at mdn for each).

Also we have to need tests for them all.

That said: better CSS support is great and I love it that you’re pushing the state here!

File: src/freenet/client/filter/HTMLFilter.java
Comment (ArneBab) [reply src/freenet/client/filter/HTMLFilter.java:1006]: OK — thank you! Supporting ruby looks like a big improvement for Asian languages.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:234]: … so yes, disabled should be useful, even when we can’t use all of the capabilities of forms.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:230]: You’re right — if they get the link from outside, it could just be different links, and if they click it on the page, it could lead anywhere in the net.

⇒ good to go. Thank you!

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:218]: A future improvement to keep pages working could be to transparently rewrite "link" to "any-link". But that may be beyond the scope of this PR (no need to fix everything in one step).

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (ArneBab) [reply src/freenet/client/filter/CSSTokenizerFilter.java:1627]: OK.

**Comment by torusrxxx** [comment]
Some more tests have been added. However, tests can only be written for expected inputs, but problems occur for unexpected inputs which cannot be tested before you find a problem! I already found two bugs that can crash the CSS validator within just a few days - very scary indeed!

**Comment by ArneBab** [comment]
Please note when you finished adding elements so I can review it again.

**Comment by torusrxxx** [comment]
Ok, I will freeze this one from now on. I do have further plans, but it seems getting code reviewed in this project is not as quickly as my first contribution being released in a new version in just a day. Hope you will have more free time in the future!

**Comment by torusrxxx** [comment]
Checking MDN for updated elements and add them here usually takes just a small amount of time that could otherwise watch a video. But I am gonna watch videos as so many things are waiting to be merged.

File: test/freenet/client/filter/CSSParserTest.java
Comment (ArneBab) [inline test/freenet/client/filter/CSSParserTest.java:42]: Does this also filter out h1:link without content?

File: test/freenet/client/filter/CSSParserTest.java
Comment (ArneBab) [inline test/freenet/client/filter/CSSParserTest.java:141]: good catch!

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:4946]: nice!

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [inline src/freenet/client/filter/ElementInfo.java:430]: why the switch to isBannedPseudoClass2 ?

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (ArneBab) [inline src/freenet/client/filter/CSSTokenizerFilter.java:1657]: good catch!

File: test/freenet/client/filter/CSSParserTest.java
Comment (ArneBab) [inline test/freenet/client/filter/CSSParserTest.java:1005]: might just not be implemented yet.

**Comment by ArneBab** [comment]
I now reviewed all the commits since my last review ( https://github.com/hyphanet/fred/pull/990#issuecomment-2425690790 ). Once you address the comments, I’ll gladly merge to next.

(no need to address "good catch" ☺)

Thank you!

File: src/freenet/client/filter/ElementInfo.java
Comment (torusrxxx) [reply src/freenet/client/filter/ElementInfo.java:430]: isBannedPseudoClass2 is a newly added function, in order to avoid `String[] split = cname.split(":");` again for just a single class and improve performance.

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java:393]: Yeah, it doesn't always work, either fails to recognize some valid forms, or throws an exception. In the end I switched to this, which passes all tests.

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (torusrxxx) [reply src/freenet/client/filter/CSSTokenizerFilter.java:4946]: There was already #830 (and more broadly #954 #955 and so on). I think it would be nice if we can get these into next as well.

File: test/freenet/client/filter/CSSParserTest.java
Comment (torusrxxx) [reply test/freenet/client/filter/CSSParserTest.java:1005]: Yeah, it's still not supported yet, but I don't like texts glowing like rainbow anyway, will leave it to future.

File: src/freenet/client/filter/ElementInfo.java
Comment (torusrxxx) [reply src/freenet/client/filter/ElementInfo.java]: dir(ltr) and dir(rtl) as pseudoclasses, added here.

File: src/freenet/client/filter/ElementInfo.java
Comment (ArneBab) [reply src/freenet/client/filter/ElementInfo.java:430]: Nice! Thank you!

File: src/freenet/client/filter/CSSTokenizerFilter.java
Comment (ArneBab) [reply src/freenet/client/filter/CSSTokenizerFilter.java:4946]: #830 needs a review and #955 needs changes from review, but #954 is ready to be merged — will merge it.

**Comment by ArneBab** [comment]
Merged — thank you!

**Pull request merged**
